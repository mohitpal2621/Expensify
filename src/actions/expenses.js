import { v4 as uuid } from "uuid";

// ADD_EXPENSE action generator
export const addExpense = ({ description='', note='', amount=0, createdAt=0 }) => ({
    type: 'ADDEXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

// REMOVE_EXPENSE action generator
export const removeExpense = (id) => ({
    type: 'REMOVEEXPENSE',
    id
});

// EDIT_EXPENSE action generator
export const editExpense = (id, updates) => ({
        type: 'EDITEXPENSE',
        id,
        updates
});