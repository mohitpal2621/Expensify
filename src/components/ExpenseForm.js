import React from "react";
import moment from "moment";
import 'react-dates/initialize'
import { SingleDatePicker } from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false,
        error: ''
    }

    onDescChange = (e) => {
        this.setState(() => (
            {description: e.target.value}
        ));
    }

    onNoteChange = (e) => {
        this.setState(() => (
            {note: e.target.value}
        ));
    } 

    onAmountChange = (e) => {
        const amt = e.target.value;
        if(!amt || /^\d+(\.\d{0,2})?$/.test(amt))
            this.setState(() => ( {amount: e.target.value} ));
    }

    onDateChange = (createdAt) => {
        if(createdAt)
            this.setState(() => ({createdAt}));
    }

    onFocusChange= ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => (
                {error: "Please provide description and amount."}
            ));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10).toFixed(2),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }

    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescChange}
                    />
                    <input 
                        type="text"
                        value={this.state.amount}
                        placeholder="Amount"
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        focused={this.state.calendarFocused}
                        onDateChange={this.onDateChange}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    {this.state.error && <p>{this.state.error}</p>}
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}