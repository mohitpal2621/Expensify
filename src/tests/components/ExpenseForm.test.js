import React from "react";
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import moment from "moment";
import ExpenseForm from '../../components/ExpenseForm';
import expenses from "../fixtures/expenses";


jest.mock('react-dates', () => ({
    SingleDatePicker: () => <div>SingleDatePicker</div>
}));

// Mock moment in ExpenseForm.test.js
jest.mock('moment', () => {
    const moment = jest.requireActual('moment');
    return (timestamp = 0) => {
        return moment(timestamp);
    };
});


// Default call with default expense
test("Should render expense form correctly", () => {
    const { container } = render(<ExpenseForm />);
    expect(container).toMatchSnapshot();
});

test("Should render expense form WITH props correctly", () => {
    const { container} = render(<ExpenseForm expense={expenses[2]} />);
    expect(container).toMatchSnapshot();
});

test("Should render ERROR for invalid form submission", () => {
    const { getByText, container } = render(<ExpenseForm />);
  
    // Take a snapshot before form submission
    expect(container).toMatchSnapshot();

    // Simulate form submission without filling in any fields
    fireEvent.submit(container.querySelector('form'));

    // Assert that the error message appears
    expect(getByText("Please provide description and amount.")).toBeInTheDocument();

    // Take a snapshot after form submission
    expect(container).toMatchSnapshot();
});

test("Should set DESCRIPTION on input change", () => {
    const { container } = render(<ExpenseForm/>);

    const description = container.querySelector(
        'input[data-testid="description-input"]'
    );

    fireEvent.change(description, {target: {value: "New CHANGE"}})

    expect(description).toHaveValue("New CHANGE");
});

test("Should set NOTE on input change", () => {
    const { container } = render(<ExpenseForm/>);

    const note = container.querySelector('textarea');

    fireEvent.change(note, {target: {value: "New CHANGE"}})

    expect(note).toHaveValue("New CHANGE");
});

test("Should set AMOUNT on VALID input change", () => {
    const { container } = render(<ExpenseForm/>);

    const amount = container.querySelector('input[data-testid="amount-input"]');

    fireEvent.change(amount, {target: {value: "6969.69"}});

    expect(amount).toHaveValue("6969.69");
});

test("Should set AMOUNT on INVALID input change", () => {
    const { container } = render(<ExpenseForm/>);

    const amount = container.querySelector('input[data-testid="amount-input"]');

    fireEvent.change(amount, {target: {value: "6969.6969"}});

    expect(amount).toHaveValue("");
});

test("Should call onSubmit PROP for valid form submission", () => {
    const onSubmitSpy = jest.fn();

    const { container } = render(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />);

    fireEvent.submit(container.querySelector('form'));
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[0].description,
        note: expenses[0].note,
        amount: parseFloat(expenses[0].amount, 10).toFixed(2),
        createdAt: expenses[0].createdAt
    });
    expect(container).toMatchSnapshot();
});
