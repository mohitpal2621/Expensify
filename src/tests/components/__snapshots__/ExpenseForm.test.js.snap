// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should call onSubmit PROP for valid form submission 1`] = `
<div>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        data-testid="description-input"
        placeholder="Description"
        type="text"
        value="Gum"
      />
      <input
        data-testid="amount-input"
        placeholder="Amount"
        type="text"
        value="140"
      />
      <div>
        SingleDatePicker
      </div>
      <textarea
        data-testid="note-textarea"
        placeholder="Add a note for your expense (optional)"
      />
      <button
        data-testid="expense-form-submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;

exports[`Should render ERROR for invalid form submission 1`] = `
<div>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        data-testid="description-input"
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        data-testid="amount-input"
        placeholder="Amount"
        type="text"
        value=""
      />
      <div>
        SingleDatePicker
      </div>
      <textarea
        data-testid="note-textarea"
        placeholder="Add a note for your expense (optional)"
      />
      <button
        data-testid="expense-form-submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;

exports[`Should render ERROR for invalid form submission 2`] = `
<div>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        data-testid="description-input"
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        data-testid="amount-input"
        placeholder="Amount"
        type="text"
        value=""
      />
      <div>
        SingleDatePicker
      </div>
      <textarea
        data-testid="note-textarea"
        placeholder="Add a note for your expense (optional)"
      />
      <p>
        Please provide description and amount.
      </p>
      <button
        data-testid="expense-form-submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;

exports[`Should render expense form WITH props correctly 1`] = `
<div>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        data-testid="description-input"
        placeholder="Description"
        type="text"
        value="Credit"
      />
      <input
        data-testid="amount-input"
        placeholder="Amount"
        type="text"
        value="113221.8"
      />
      <div>
        SingleDatePicker
      </div>
      <textarea
        data-testid="note-textarea"
        placeholder="Add a note for your expense (optional)"
      >
        Credit card bills
      </textarea>
      <button
        data-testid="expense-form-submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;

exports[`Should render expense form correctly 1`] = `
<div>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        data-testid="description-input"
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        data-testid="amount-input"
        placeholder="Amount"
        type="text"
        value=""
      />
      <div>
        SingleDatePicker
      </div>
      <textarea
        data-testid="note-textarea"
        placeholder="Add a note for your expense (optional)"
      />
      <button
        data-testid="expense-form-submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;
